src/Microsoft.AspNet.Mvc.Core/Filters/ResponseCacheAttribute.cs

```diff
@@ -38,97 +38,85 @@ public int Duration
         /// Ignores the "duration" parameter.
         /// </summary>
-        public bool NoStore { get; set; }
+        public bool NoStore
+        {
+            get
+            {
+                return _noStore ?? false;
+            }
+            set
+            {
+                _noStore = value;
+            }
+        }
 
         /// <summary>
         /// Gets or sets the value for the Vary response header.
         /// </summary>
         public string VaryByHeader { get; set; }
 
-        // <inheritdoc />
-        public override void OnActionExecuting([NotNull] ActionExecutingContext context)
-        {
-            // If there are more filters which can override the values written by this filter,
-            // then skip execution of this filter.
-            if (IsOverridden(context))
-            {
-                return;
-            }
-
-            var headers = context.HttpContext.Response.Headers;
-
-            // Clear all headers
-            headers.Remove("Vary");
-            headers.Remove("Cache-control");
-            headers.Remove("Pragma");
+        /// <summary>
+        /// Gets or sets the value of the Cache Profile.
+        /// </summary>
+        public string CacheProfileName { get; set; }
 
-            if (!string.IsNullOrEmpty(VaryByHeader))
-            {
-                headers.Set("Vary", VaryByHeader);
-            }
+        public IFilter CreateInstance([NotNull] IServiceProvider serviceProvider)
+        {
+            var optionsAccessor = serviceProvider.GetRequiredService<IOptions<MvcOptions>>();
 
-            if (NoStore)
+            CacheProfile selectedProfile = null;
+            if (!string.IsNullOrEmpty(CacheProfileName))
             {
-                headers.Set("Cache-control", "no-store");
+                selectedProfile = optionsAccessor.Options.CacheProfiles.FirstOrDefault(
+                    p => string.Equals(p.Name, CacheProfileName, StringComparison.OrdinalIgnoreCase));
 
-                // Cache-control: no-store, no-cache is valid.
-                if (Location == ResponseCacheLocation.None)
+                if (selectedProfile == null)
                 {
-                    headers.Append("Cache-control", "no-cache");
-                    headers.Set("Pragma", "no-cache");
+                    throw new InvalidOperationException(Resources.FormatCacheProfileNotFound(CacheProfileName));
                 }
             }
-            else
+
+            // If the ResponseCacheAttribute parameters are set,
+            // then it must override the values from the Cache Profile.
+            // The below expression first checks if the duration is set by the attribute's parameter.
+            // If absent, it checks the selected cache profile (Note: There can be no cache profile as well)
+            // The same is the case for other properties.
+            _duration = _duration ?? selectedProfile?.Duration;
```

Comment:
I did not pass `CacheProfile` because it contains nullable values and the `ResponseCacheFilter` does not. Now when `ResponseCacheAttribute` takes nullable values we must write code to convert these nullables to their defaults. This code will now be duplicated in `ResponseCacheFilter` and `ResponseCacheAttribute`.